File: myLeetCode/Binary Tree/二叉树的遍历方式/145. Binary Tree Postorder Traversal.cpp
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        preorder(root, res);
        return res;
    }

private:
    void preorder(TreeNode* node, vector<int> &res) {
        if (node == NULL) return;
        res.push_back(node->val);
        preorder(node->left, res);
        preorder(node->right, res);
    }
};

int main() {
    Solution sol;
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    vector<int> res = sol.preorderTraversal(root);
    for (int x : res) {
        cout << x << " ";
    }
    cout << endl;
    return 0;
}

File: C#/myLeetCode/Binary_Tree/Traversal mode of Binary tree/iteration/145. Binary Tree Postorder Traversal.cs
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/

using System;
using System.Collections.Generic;

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    public IList<int> PostorderTraversal(TreeNode root) {
        IList<int> result = new List<int>();
        if (root == null) return result;

        Stack<TreeNode> stack1 = new Stack<TreeNode>();
        Stack<TreeNode> stack2 = new Stack<TreeNode>();
        stack1.Push(root);

        while (stack1.Count > 0) {
            TreeNode currentNode = stack1.Pop();
            stack2.Push(currentNode);

            if (currentNode.left != null) {
                stack1.Push(currentNode.left);
            }
            if (currentNode.right != null) {
                stack1.Push(currentNode.right);
            }
        }

        while (stack2.Count > 0) {
            result.Add(stack2.Pop().val);
        }

        return result;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        TreeNode tree = new TreeNode(1) {
            right = new TreeNode(2) {
                left = new TreeNode(3)
            }
        };

        IList<int> postorderTraversal = solution.PostorderTraversal(tree);

        Console.WriteLine("Postorder Traversal:");
        foreach (int val in postorderTraversal) {
            Console.Write(val + " ");
        }
    }
}
